### PR TITLE
REF: _cython_agg_blocks

### PR DESCRIPTION
**Pull request by jbrockmendel**
This is effectively an alternative approach to #31616, trying to use a 1-pass approach. 

This does _not_ address the uniqueness problem discussed in #31735.

cc @TomAugspurger @WillAyd

### CODE REVIEW

File: pandas/core/groupby/generic.py [hunk 1]
```diff
@@ -1063,58 +1087,32 @@ def _cython_agg_blocks(
                     deleted_items.append(locs)
                     continue
                 else:
+                    if isinstance(result, Series):
```
Comment (WillAyd) [inline pandas/core/groupby/generic.py]: Hmm why is this required here?

File: pandas/core/groupby/generic.py [hunk 2]
```diff
@@ -1063,58 +1087,32 @@ def _cython_agg_blocks(
...
-                    assert result.ndim == 1 or result.shape[0] == 1
-                    try:
-                        # Cast back if feasible
-                        result = type(block.values)._from_sequence(
-                            result.ravel(), dtype=block.values.dtype
-                        )
-                    except ValueError:
-                        # reshape to be valid for non-Extension Block
-                        result = result.reshape(1, -1)
+                    for i, col in enumerate(result.columns):
```
Comment (WillAyd) [inline pandas/core/groupby/generic.py]: Perhaps misinterpreting but does this not have a huge performance penalty? IIUC tried column iteration back in #28782 but performance took a pretty big hit in doing so

File: pandas/core/groupby/generic.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/groupby/generic.py]: we have test cases that get here with a Series and without this the `cast` just below here is inaccurate

File: pandas/core/groupby/generic.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/groupby/generic.py]: the actual aggregation isn't being done columnwise, just re-assembling new blocks

File: pandas/core/groupby/generic.py (hunk 1 above)
Comment (WillAyd) [reply pandas/core/groupby/generic.py]: Hmm that's unfortunate. Don't want to go too far down the typing road on this one but maybe should be using `assert ...` instead of `cast`

Comment (WillAyd) [reply pandas/core/groupby/generic.py]: cc @simonjayhawkins

File: pandas/core/groupby/generic.py [hunk 3]
```diff
@@ -1022,9 +1022,32 @@ def _cython_agg_blocks(
         agg_blocks: List[Block] = []
         new_items: List[np.ndarray] = []
         deleted_items: List[np.ndarray] = []
-        # Some object-dtype blocks might be split into List[Block[T], Block[U]]
-        split_items: List[np.ndarray] = []
-        split_frames: List[DataFrame] = []
+
+        def _recast_result(result, values):
```
Comment (jreback) [inline pandas/core/groupby/generic.py]: can you move this somewhere else, pandas.core.dtypes.cast maybe

**Review by jreback — CHANGES_REQUESTED** [review]
this is still really pretty gnarly. do you have time to really refactor this? e.g. move to internals & then cleanup.

File: pandas/core/groupby/generic.py (hunk 1 above)
Comment (simonjayhawkins) [reply pandas/core/groupby/generic.py]: hmm, `reveal_type(result)` entering this block is object. probably need to take control of the type of result before this. does adding type annotations to  `_recast_result` help?

Comment (simonjayhawkins) [reply pandas/core/groupby/generic.py]: it looks like since `aggregate` in `SelectionMixin` does not have type annotations, the return type of `s.aggregate(lambda x: alt(x, axis=self.axis))` is `Any` and hence as result is object no type narrowing occurs. reveal_type(s) is 'pandas.core.groupby.groupby.GroupBy' so it's just `aggregate` that needs the annotations.

**Comment by jbrockmendel** [comment]
closing to clear the queue
